packages/sdk/src/cross-chain-messenger.ts

```diff
@@ -1279,29 +1311,100 @@ export class CrossChainMessenger {
-          match = {
-            ...game,
-            l2BlockNumber: blockNumber,
+        try {
+          const [blockNumber] = ethers.utils.defaultAbiCoder.decode(
+            ['uint256'],
+            game.extraData
+          )
+          if (blockNumber.gte(resolved.blockNumber)) {
+            matches.push({
+              ...game,
+              l2BlockNumber: blockNumber,
+            })
           }
-          break
+        } catch (err) {
+          // If we can't decode the extra data then we just skip this game.
+          continue
         }
       }
 
-      // TODO: It would be more correct here to actually verify the proposal since proposals are
-      // not guaranteed to be correct. proveMessage will actually do this verification for us but
-      // there's a devex edge case where this message appears to give back a valid proposal that
-      // ends up reverting inside of proveMessage. At least this is safe for users but not ideal
-      // for developers and we should work out the simplest way to fix it. Main blocker is that
-      // verifying the proposal may require access to an archive node.
+      // Shuffle the list of matches. We shuffle here to avoid potential DoS vectors where the
+      // latest games are all invalid and the SDK would be forced to make a bunch of archive calls.
+      for (let i = matches.length - 1; i > 0; i--) {
+        const j = Math.floor(Math.random() * (i + 1))
+        ;[matches[i], matches[j]] = [matches[j], matches[i]]
+      }
+
+      // Now we verify the proposals in the matches array.
+      let match: any
+      for (const option of matches) {
+        // Use the cache if we can.
+        const cached = this._outputCache.find((other) => {
+          return other.root === option.rootClaim
+        })
+
+        // Skip if we can use the cached.
+        if (cached) {
+          if (cached.valid) {
+            match = option
+            break
+          } else {
+            continue
+          }
+        }
+
+        // If the cache ever gets to 10k elements, clear out the first half. Works well enough
+        // since the cache will generally tend to be used in a FIFO manner.
+        if (this._outputCache.length > 10000) {
+          this._outputCache = this._outputCache.slice(5000)
+        }
+
+        // We didn't hit the cache so we're going to have to do the work.
+        try {
+          // Grab the raw block JSON (need state root).
+          const block = await (
+            this.l2Provider as ethers.providers.JsonRpcProvider
+          ).send('eth_getBlockByNumber', [
+            toRpcHexString(option.l2BlockNumber),
+            false,
+          ])
+
+          // Grab the storage proof.
+          const proof = await makeStateTrieProof(
+            this.l2Provider as ethers.providers.JsonRpcProvider,
+            option.l2BlockNumber,
+            this.contracts.l2.OVM_L2ToL1MessagePasser.address,
+            ethers.constants.HashZero
+          )
```

Comment:
Small optimization to avoid an unneeded round trip 

```suggestion
          // I think this is necessary in case user passes in a fallback provider?
          if (!('send' in this.l2Provider)) throw new Error('Unable to process because l2Provider isn't a valid JsonRpcProvider')
          
          // can fetch both of these at same time
          const [block, proof] = await Promise.all([
          (
            this.l2Provider as ethers.providers.JsonRpcProvider
          ).send('eth_getBlockByNumber', [
            toRpcHexString(option.l2BlockNumber),
            false,
          ]),
          // Grab the storage proof.
          makeStateTrieProof(
            this.l2Provider as ethers.providers.JsonRpcProvider,
            option.l2BlockNumber,
            this.contracts.l2.OVM_L2ToL1MessagePasser.address,
            ethers.constants.HashZero
          )
          ])
```